### PR TITLE
Pdo

### PR DESCRIPTION
**Pull request by jan-pavlovsky**
class PDO complete besides transactions

PDOStatement::
execute() redone
bindValue()
bindParam() - TODO references to PhpAlias problem

### CODE REVIEW

File: src/PDO/Peachpie.Library.PDO/PDOStatement.cs [hunk 1]
```diff
@@ -32,25 +32,37 @@ public class PDOStatement : IPDOStatement, IDisposable
         private bool m_namedAttr = false;
         private Dictionary<string, string> m_namedPlaceholders;
         private List<String> m_positionalPlaceholders;
+
+        private bool m_hasParamsBounded = false;
```
Comment (jakubmisek) [inline src/PDO/Peachpie.Library.PDO/PDOStatement.cs]: can we just remove this field and replace it with a property like
`bool HasParamsBound => m_boundedParams != null;`

File: src/PDO/Peachpie.Library.PDO/PDOStatement.cs [hunk 2]
```diff
@@ -32,25 +32,37 @@ public class PDOStatement : IPDOStatement, IDisposable
...
-        /// <param name="pdo">The pdo.</param>
+        /// <param name="pdo">The PDO statement is created for.</param>
         /// <param name="statement">The statement.</param>
         /// <param name="driver_options">The driver options.</param>
         internal PDOStatement(PDO pdo, string statement, PhpArray driver_options)
         {
             this.m_pdo = pdo;
             this.m_stmt = statement;
             this.m_options = driver_options ?? PhpArray.Empty;
+            this.m_hasParamsBounded = false;
```
Comment (jakubmisek) [inline src/PDO/Peachpie.Library.PDO/PDOStatement.cs]: unnecessary

File: src/PDO/Peachpie.Library.PDO/PDOStatement.cs [hunk 3]
```diff
@@ -63,6 +75,7 @@ protected PDOStatement()
             m_stmt = null;
             m_options = PhpArray.Empty;
             m_cmd = null;
+            m_hasParamsBounded = false;
```
Comment (jakubmisek) [inline src/PDO/Peachpie.Library.PDO/PDOStatement.cs]: unnecessary

File: src/PDO/Peachpie.Library.PDO/PDOStatement.cs [hunk 4]
```diff
@@ -243,31 +263,259 @@ public bool bindColumn(PhpValue colum, ref PhpValue param, int? type = default(i
         }
 
         /// <inheritDoc />
-        public bool bindParam(PhpValue parameter, ref PhpValue variable, int data_type = 2, int? length = default(int?), PhpValue? driver_options = default(PhpValue?))
+        public bool bindParam(PhpValue parameter, PhpAlias variable, PDO.PARAM data_type = PDO.PARAM.PARAM_STR, int length = default(int), PhpValue driver_options = default(PhpValue))
```
Comment (jakubmisek) [inline src/PDO/Peachpie.Library.PDO/PDOStatement.cs]: `default(int)` is `0`, is it alright?

Comment (jan-pavlovsky) [reply src/PDO/Peachpie.Library.PDO/PDOStatement.cs]: It is not used anywhere at the moment, as the Statement option is not implemented so far. However agreed that -1 is a better option.

**Comment by jan-pavlovsky** [comment]
Thank you, agreed and changed.

**Pull request merged**